File: 2580. Count Ways to Group Overlapping Ranges/Solution.java
class Solution {
    public int countWays(int[][] ranges) {
        Arrays.sort(ranges, (a, b) -> {
            if (a[0] == b[0]) return b[1] - a[1];
            return a[0] - b[0];
        });
        int upper = -1, count = 0;
        for (int[] r : ranges) {
            if (r[0] <= upper) {
                upper = Math.max(upper, r[1]);
            } else {
                count++;
                upper = r[1];
            }
        }
        long res = 1, MOD = 1000000007;
        for (int i = 0; i < count; i++) {
            res = (res * 2) % MOD;
        }
        return (int) res;
    }
}

File: 2580. Count Ways to Group Overlapping Ranges/Solution.py
class Solution:
    def countWays(self, ranges: List[List[int]]) -> int:
        ranges = sorted(ranges, key = itemgetter(0, 1))
        upper, count, res, mod = -1, 0, 1, 10 ** 9 + 7
        for left, right in ranges:
            if left <= upper:
                upper = max(upper, right)
            else:
                count += 1
                upper = right
        for i in range(count):
            res = (res * 2) % mod
        return res